### ISSUE TITLE
Question about ppo_epochs

### ISSUE THREAD

**Issue by theoqian** [BytedTsinghua-SIA/DAPO#11]
In the DAPO script https://github.com/volcengine/verl/blob/gm-tyx/puffin/main/recipe/dapo/run_dapo_qwen2.5_32b.sh, ppo_epochs is not set which means ppo_epochs takes the default value 1. However, if ppo_epochs equals to 1, then clip_ratio_low and clip_ratio_high won't take effect since the ratio is always 1.0 for the first ppo_epochs. So what's the actual value of ppo_epochs?

**Comment by huiyeruzhou** [comment]
There might be some misunderstandings here. Let's review several key hyperparameters in the reinforcement learning algorithm together:
1. The old actor generates a batch of samples, and the number of these samples is determined by `data.train_batch_size`.
2. A part of the samples are selected for gradient update, and the number of these samples is determined by `actor_rollout_ref.actor.ppo_mini_batch_size`.
3. During the actual calculation, considering the limitation of GPU memory, only a small number of samples are calculated each time. The number of these samples is `actor_rollout_ref.actor.ppo_micro_batch_size`, and the method of gradient accumulation is applied.
4. Sometimes, in order to achieve off - policy update, we reuse the generated samples multiple times, and the number of reuse times is determined by `ppo_epochs`.


> However, if ppo_epochs equals to 1, then clip_ratio_low and clip_ratio_high won't take effect since the ratio is always 1.0 for the first ppo_epochs. 

It should be noted here that the factor causing importance sampling is not only `ppo_epochs`. If `train_batch_size` is not equal to `ppo_mini_batch_size`, there is already a deviation between the optimized actor and the sampling actor, thus clipping occurs. Therefore, the value of `ppo_epochs` is 1, and the place where clipping takes effect is when `train_batch_size = 512` and `ppo_mini_batch_size = 32`. This means that in addition to the first update, clipping is involved in the remaining 7 updates.

**Comment by huiyeruzhou** [comment]
Hi!@glowwormX Your understanding seems correct to me. Based solely on the figure, GRPO does appear to indicate that minibs=bs and ppo_epochs>1.

However, there are several reasons why minibatches (minibs) are typically used for off-policy updates instead of increasing ppo_epochs:

Established Convention: Following existing practices is ideal for comparability and leveraging prior empirical knowledge, as this approach has become standard in recent RL experiments.

Performance Concerns: Early work like [PPG](https://arxiv.org/abs/2009.04416,) along with other academic and open-source implementations of PPO, has shown that sample reuse (ppo_epochs > 1) for the actor policy updating can lead to significant performance degradation.

Computational Trade-offs: Since losses are averaged within the same batch, larger batch sizes only marginally improve update stability while incurring substantially higher computational costs—in addition to the performance issues mentioned in point 2.

In summary, setting ppo_epochs=1 and handling off-policy updates through minibatches is a common practice with theoretical and empirical evidence.

I’m also curious about the effects of ppo_epochs>1 and would love to discuss this further if you’ve run any experiments on it！

**Comment by glowwormX** [comment]
@huiyeruzhou In my experiment, ppo_epochs>1 is very similar to what you said, with low training efficiency and poor performance. And in deepseek's paper, I seem to see μ=1. What I am concerned about is minibs=bs, and clip is not used. Whether the performance of `minibs=bs` is better than that of `minibs<bs`. If dynamic sampling is required, we can supplement the filtered prompt through dynamic filtering and new prompt of dynamic sample.
